sql-plugin/src/main/330db/scala/org/apache/spark/sql/rapids/execution/GpuExecutorBroadcastHelper.scala

```diff
@@ -54,23 +67,40 @@ object GpuExecutorBroadcastHelper extends Arm {
    * Given an RDD of ColumnarBatch containing broadcast data from a shuffle, get the 
    * fully coalesced ColumnarBatch loaded on to the GPU.
    *
-   * @param buildRelation - the executor broadcast produced by a shuffle exchange
+   * This method uses getSingleBatchWithVerification to handle the empty relation case 
+   * and the unexpected scenario of receiving more than a single batch
+   *
+   * @param shuffleData - the shuffle data from the executor for the "executor broadcast"
    * @param buildSchema - the schema expected for the output
    * @param buildOutput - the output attributes expected in case we receieve an empty RDD
    * @param metricsMap - metrics to populate while the shuffle coalesce iterator loads 
    *                     data from I/O to the device
    * @param targetSize - the target batch size in bytes for loading data on to the GPU 
    */
   def getExecutorBroadcastBatch(
-      buildRelation: RDD[ColumnarBatch], 
+      shuffleData: RDD[ColumnarBatch], 
       buildSchema: StructType,
       buildOutput: Seq[Attribute],
       metricsMap: Map[String, GpuMetric],
       targetSize: Long): ColumnarBatch = {
 
-    val it = shuffleCoalesceIterator(buildRelation, buildSchema, metricsMap, targetSize)
+    val it = shuffleCoalesceIterator(shuffleData, buildSchema, metricsMap, targetSize)
     // In a broadcast hash join, right now the GPU expects a single batch
     ConcatAndConsumeAll.getSingleBatchWithVerification(it, buildOutput)
   }
 
+  /**
+   * Given an RDD of ColumnarBatch containing broadcast data from a shuffle, get the 
+   * number of rows that the received batch contains
+   */
+  def getExecutorBroadcastBatchNumRows(shuffleData: RDD[ColumnarBatch]): Int = {
+    val it = shuffleDataIterator(shuffleData)
+    if (it.hasNext) {
+      val batch = it.next
+      batch.numRows
+    } else {
+      0
+    }
```

Comment:
This seems to have a couple of issues.  We're iterating the RDD more than once?  That will cause it to re-execute parts of the plan, probably causing it to re-fetch the shuffle data.  This also doesn't close the batch that was received.